internal/modelcontroller/adapters.go

```diff
@@ -0,0 +1,210 @@
+package modelcontroller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	v1 "github.com/substratusai/kubeai/api/v1"
+	"github.com/substratusai/kubeai/internal/k8sutils"
+	"github.com/substratusai/kubeai/internal/vllmclient"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	loaderContainerName = "loader"
+)
+
+func (r *ModelReconciler) reconcileAdapters(ctx context.Context, pods []*corev1.Pod, adapters []v1.Adapter) error {
+	type reconcileParam struct {
+		pod         *corev1.Pod
+		toEnsure    []v1.Adapter
+		toRemoveIDs []string
+		engine      string
+	}
+	var reconcileList []reconcileParam
+
+	for _, pod := range pods {
+		param := reconcileParam{
+			pod: pod,
+		}
+
+		if pod.Labels == nil {
+			continue
+		}
+		switch pod.Labels[appKubernetesIOName] {
+		case strings.ToLower(v1.VLLMEngine):
+			param.engine = v1.VLLMEngine
+		default:
+			continue
+		}
+
+		deletionCandidates := getLabelledAdapters(pod)
+
+		for _, adapter := range adapters {
+			if k8sutils.GetLabel(pod, v1.PodAdapterLabel(adapter.Name)) != k8sutils.StringHash(adapter.URL) {
```

Comment:
Feel free to ignore this. Just writing down notes as I get through the review.

I don't get  this comparison. This should always be false if we're comparing regular string against a string hash.

This is from docs  "// StringHash returns a hash value calculated from the input string."

So shouldn't this be this instead:
```
if k8sutils.GetLabel(pod, v1.PodAdapterLabel(adapter.Name)) != adapter.URL {
```

Maybe GetLabel returns a hash as well? then it would make sense.